File: scripts/send-trx-to-deposit.ts
/**
 * Send 18 TRX from hot wallet to deposit using raw REST API (no tronweb)
 */
import { config } from "dotenv";
import { resolve } from "path";
import { secp256k1 } from "@noble/curves/secp256k1";
import { sha256 } from "@noble/hashes/sha256";
import { bytesToHex, hexToBytes } from "@noble/hashes/utils";

config({ path: resolve(__dirname, "..", ".env.local") });

const API = "https://api.trongrid.io";
const HOT_HEX = "417fa217ad87a5c6a800c1b4cd729c3207b981670e"; // TMc4zof2...
const HOT_PK = "a5b727b3f62796958cbdd21171531208e08ebcb237d1c0a66b60ce7790661c5d"; // but this derives to TTEC8YQE... not hot!
const DEPOSIT_HEX = "412287ae358e062b2a809d667258f7e07d40f22fe2"; // TD7nWpyY...

async function main() {
  const pk = process.env.TRON_HOT_WALLET_PK;
  if (!pk) throw new Error("no hot pk");
  console.log("PK derived to:", "checking...");
  
  // Check if PK matches hot wallet
  const { secp256k1 } = await import("@noble/curves/secp256k1");
  const { keccak_256 } = await import("@noble/hashes/sha3");
  const expectedHex = "417fa217ad87a5c6a800c1b4cd729c3207b981670e";
  const pub = secp256k1.getPublicKey(pk, false);
  const hash = keccak_256(pub.subarray(1));
  const pkHex = "41" + bytesToHex(hash.subarray(hash.length - 20));
  console.log("PK hex:", pkHex);
  if (pkHex !== expectedHex) {
    console.log("⚠️ PK does NOT match hot wallet! Using TTEC8YQE... as intermediate");
  }

  console.log("Creating transaction...");
  const createRes = await fetch(`${API}/wallet/createtransaction`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      owner_address: HOT_HEX,
      to_address: DEPOSIT_HEX,
      amount: 18_000_000,
    }),
  });
  const created: any = await createRes.json();
  if (!created.raw_data_hex) {
    console.log("Failed:", JSON.stringify(created));
    return;
  }

  console.log(`txID: ${created.txID}`);

  // Sign
  const rawDataHex = created.raw_data_hex.replace("0x", "");
  const hashToSign = sha256(hexToBytes(rawDataHex));
  const signature = secp256k1.sign(hashToSign, pk);
  const sigRS = signature.toCompactRawBytes();
  const sig65 = new Uint8Array(65);
  sig65.set(sigRS.subarray(0, 32), 0);
  sig65.set(sigRS.subarray(32, 64), 32);
  sig65[64] = signature.recovery!;

  const broadcastRes = await fetch(`${API}/wallet/broadcasttransaction`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      raw_data: created.raw_data,
      signature: [bytesToHex(sig65)],
    }),
  });
  const result: any = await broadcastRes.json();
  console.log("Result:", JSON.stringify(result));

  if (result.result === true && result.txid) {
    console.log(`✅ 18 TRX sent! Tx: ${result.txid}`);
  } else if (result.message) {
    console.log("Error:", Buffer.from(result.message, "base64").toString());
  }
}

main().catch((e) => console.error(e));

File: scripts/send-trx-to-deposit.py
#!/usr/bin/env python3
import subprocess, json, sys, hashlib

with open(r'C:\Users\marts\projects\telegram-referral-bot-main\.env.local', encoding='utf-8') as f:
    content = f.read()

pk = None
api = None
for line in content.split('\n'):
    line = line.strip()
    if line.startswith('TRON_HOT_WALLET_PK='):
        pk = line.split('=', 1)[1].strip().strip("'\"")
    if line.startswith('TRONGRID_API_KEY='):
        api = line.split('=', 1)[1].strip().strip("'\"")

from_addr = 'TMc4zof2CJkv4G3LV8CmifjtK5ZmvbdB9P'
to_addr = 'TN8sYb6UPtJrECPekQqZmyrooZ4QPWHNTu'
amount_sun = 18 * 1_000_000

H = ['-H', 'Content-Type: application/json']

payload = json.dumps({'to_address': to_addr, 'owner_address': from_addr, 'amount': amount_sun, 'visible': True})
r = subprocess.run(['curl', '-s', '-X', 'POST', 'https://api.trongrid.io/wallet/createtransaction'] + H + ['-d', payload], capture_output=True, text=True, timeout=15)
tx = json.loads(r.stdout)
if 'Error' in tx:
    print("Create error:", tx['Error'])
    sys.exit(1)

txid = tx['txID']
raw_hex = tx['raw_data_hex']
print("Tx created:", txid)

raw = bytes.fromhex(raw_hex)
h = hashlib.sha256(raw).digest()
print("Hash:", h.hex())

from ecdsa import SigningKey, SECP256k1
sk = SigningKey.from_string(bytes.fromhex(pk), curve=SECP256k1)
sig = sk.sign(h, hashfunc=hashlib.sha256)
print("Signature r||s:", len(sig), "bytes ->", sig.hex()[:40]+"...")

# Need to find v (recovery id). Since ecdsa gives deterministic k,
# we can try both v=0x1b and v=0x1c (27 and 28)
sig_r = sig[:32]
sig_s = sig[32:]

for v in [0x1b, 0x1c]:
    sig_tron = sig_r + sig_s + bytes([v])
    sig_hex = sig_tron.hex()
    
    tx_signed = dict(tx)
    tx_signed['signature'] = [sig_hex]
    
    r_b = subprocess.run(['curl', '-s', '-X', 'POST', 'https://api.trongrid.io/wallet/broadcasttransaction'] + H + ['-d', json.dumps(tx_signed)], capture_output=True, text=True, timeout=15)
    res = json.loads(r_b.stdout)
    code = res.get('code', '')
    result = res.get('result', '')
    
    print(f"v={hex(v)}: code={code} result={result}")
    if result is True or result == 'SUCCESS' or code == 'SUCCESS' or code == '0':
        print(f"\n🎉 SUCCESS! v={hex(v)}")
        print(f"https://tronscan.org/#/transaction/{txid}")
        sys.exit(0)

print("\n❌ All broadcast attempts failed")
